rcgen/src/string_types.rs

```diff
@@ -0,0 +1,317 @@
+		&self.0
+	}
+}
+
+impl TryFrom<&str> for PrintableString {
+	type Error = Error;
+
+	fn try_from(input: &str) -> Result<Self, Error> {
+		input.to_string().try_into()
+	}
+}
+
+impl TryFrom<String> for PrintableString {
+	type Error = Error;
+
+	fn try_from(value: String) -> Result<Self, Self::Error> {
+		for &c in value.as_bytes() {
+			match c {
+				b'A'..=b'Z'
+				| b'a'..=b'z'
+				| b'0'..=b'9'
+				| b' '
+				| b'\''
+				| b'('
+				| b')'
+				| b'+'
+				| b','
+				| b'-'
+				| b'.'
+				| b'/'
+				| b':'
+				| b'='
+				| b'?' => (),
+				_ => return Err(Error::InvalidPrintableString),
+			}
+		}
+		Ok(Self(value))
+	}
+}
+
+impl FromStr for PrintableString {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		s.try_into()
+	}
+}
+
+/// ASN.1 `IA5String` type.
+///
+/// Supports the [International Alphabet No. 5 (IA5)] character encoding, i.e.
+/// the lower 128 characters of the ASCII alphabet. (Note: IA5 is now
+/// technically known as the International Reference Alphabet or IRA as
+/// specified in the ITU-T's T.50 recommendation).
+///
+/// For UTF-8, use [`String`][`std::string::String`].
+///
+/// [International Alphabet No. 5 (IA5)]: https://en.wikipedia.org/wiki/T.50_%28standard%2
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct Ia5String(String);
+
+impl Ia5String {
+	/// Extracts a string slice containing the entire `Ia5String`.
+	pub fn as_str(&self) -> &str {
+		&self.0
+	}
+}
+
+impl TryFrom<&str> for Ia5String {
+	type Error = Error;
+
+	fn try_from(input: &str) -> Result<Self, Error> {
+		input.to_string().try_into()
+	}
+}
+
+impl TryFrom<String> for Ia5String {
+	type Error = Error;
+
+	fn try_from(input: String) -> Result<Self, Error> {
+		// Validate all characters are within IA5String's allowed set
+		if input.as_bytes().iter().any(|&c| c > 0x7F) {
+			return Err(Error::InvalidPrintableString);
+		}
+		Ok(Self(input.to_string()))
+	}
+}
+
+impl FromStr for Ia5String {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		s.try_into()
+	}
+}
+
+/// ASN.1 `TeletexString` type.
+///
+/// # Supported characters
+///
+/// The standard defines a complex character set allowed in this type. However, quoting the ASN.1
+/// mailing list, "a sizable volume of software in the world treats TeletexString (T61String) as a
+/// simple 8-bit string with mostly Windows Latin 1 (superset of iso-8859-1) encoding".
+///
+/// `TeletexString` is included for backward compatibility, [RFC 5280] say it
+/// SHOULD NOT be used for certificates for new subjects.
+///
+/// [RFC 5280]: https://datatracker.ietf.org/doc/html/rfc5280#page-25
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct TeletexString(String);
+
+impl TeletexString {
+	/// Extracts a string slice containing the entire `TeletexString`.
+	pub fn as_str(&self) -> &str {
+		&self.0
+	}
+
+	/// Extracts a string slice containing the entire `PrintableString`.
+	pub fn as_bytes(&self) -> &[u8] {
```

Comment:
I think this comment isn't accurate: it's not a string slice and it's not a `PrintableString` :-)